sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -287,3 +289,152 @@ case class ArrayContains(left: Expression, right: Expression)
+    })
+  }
+
+  private def genCodeForNumberOfElements(
+    ctx: CodegenContext,
+    elements: Seq[String]
+  ) : (String, String) = {
+    val variableName = ctx.freshName("numElements")
+    val code = elements
+      .map(el => s"$variableName += $el.numElements();")
+      .foldLeft( s"int $variableName = 0;")((acc, s) => acc + "\n" + s)
+    (code, variableName)
+  }
+
+  private def genCodeForConcatOfPrimitiveElements(
+    ctx: CodegenContext,
+    elementType: DataType,
+    elements: Seq[String],
+    arrayDataName: String
+  ): String = {
+    val arrayName = ctx.freshName("array")
+    val arraySizeName = ctx.freshName("size")
+    val counter = ctx.freshName("counter")
+    val tempArrayDataName = ctx.freshName("tempArrayData")
+
+    val (numElemCode, numElemName) = genCodeForNumberOfElements(ctx, elements)
+
+    val unsafeArraySizeInBytes = s"""
+      |int $arraySizeName = UnsafeArrayData.calculateHeaderPortionInBytes($numElemName) +
+      |${classOf[ByteArrayMethods].getName}.roundNumberOfBytesToNearestWord(
+      |${elementType.defaultSize} * $numElemName
+      |);
+      """.stripMargin
+    val baseOffset = Platform.BYTE_ARRAY_OFFSET
+
+    val primitiveValueTypeName = CodeGenerator.primitiveTypeName(elementType)
+    val assignments = elements.map { el =>
+      s"""
+        |for(int z = 0; z < $el.numElements(); z++) {
+        | if($el.isNullAt(z)) {
+        |   $tempArrayDataName.setNullAt($counter);
+        | } else {
+        |   $tempArrayDataName.set$primitiveValueTypeName(
+        |     $counter,
+        |     $el.get$primitiveValueTypeName(z)
+        |   );
+        | }
+        | $counter++;
+        |}
+        """.stripMargin
+    }.mkString("\n")
+
+    s"""
+      |$numElemCode
+      |$unsafeArraySizeInBytes
+      |byte[] $arrayName = new byte[$arraySizeName];
+      |UnsafeArrayData $tempArrayDataName = new UnsafeArrayData();
+      |Platform.putLong($arrayName, $baseOffset, $numElemName);
+      |$tempArrayDataName.pointTo($arrayName, $baseOffset, $arraySizeName);
+      |int $counter = 0;
+      |$assignments
+      |$arrayDataName = $tempArrayDataName;
+    """.stripMargin
+
+  }
+
+  private def genCodeForConcatOfComplexElements(
+   ctx: CodegenContext,
+   elements: Seq[String],
+   arrayDataName: String
+  ): String = {
+    val genericArrayClass = classOf[GenericArrayData].getName
+    val arrayName = ctx.freshName("arrayObject")
+    val counter = ctx.freshName("counter")
+    val (numElemCode, numElemName) = genCodeForNumberOfElements(ctx, elements)
+
+    val assignments = elements.map { el =>
+      s"""
+        |for(int z = 0; z < $el.numElements(); z++) {
+        |  $arrayName[$counter] = $el.array()[z];
+        |  $counter++;
+        |}
+     """.stripMargin
+    }.mkString("\n")
+
+    s"""
+      |$numElemCode
+      |Object[] $arrayName = new Object[$numElemName];
+      |int $counter = 0;
+      |$assignments
+      |$arrayDataName = new $genericArrayClass($arrayName);
```

Comment:
+1, can we reuse the `UnsafeArrayWriter` logic for this case?

Reply:
Really like this idea! I think it would require moving the complex type insertion logic from `InterprettedUnsafeProjection` directly to `UnsafeDataWriter` and introduce in that way write methods for complex type fields. I'm not sure whether this big refactoring task is still in the scope of this PR.

Also see that we could improve codeGen of `CreateArray` in the same way.

Reply:
You couldn't use `UnsafeArrayData` in the complex case?

Reply:
Yeah, currently there are no `write` methods on `UnsafeArrayWriter` or `set` methods on `UnsafeArrayData` that we could leverage for complex types. In theory, we could follow the same approach as in `InterprettedUnsafeProjection` and each complex type to a byte array and subsequently insert the produced byte array into the target `UnsafeArrayData`. Since this logic could be utilized from more places (e.g. `CreateArray`), it should be encapsulated into `UnsafeArrayWriter` or  `UnsafeArrayData` at first. What do you think?